File: Chapter16/SampleViewAnimation/Java/SampleViewAnimation/app/src/main/java/com/ai/android/ExerciseSystemIntents/ViewAnimation.java
package com.ai.android.ExerciseSystemIntents;

import android.graphics.Matrix;
import android.util.Log;
import android.view.animation.Animation;
import android.view.animation.LinearInterpolator;
import android.view.animation.Transformation;

public class ViewAnimation extends Animation {

    private static final String TAG = "ViewAnimation";

    private float centerX;
    private float centerY;

    public ViewAnimation(float cx, float cy) {
        centerX = cx;
        centerY = cy;
    }

    @Override
    public void initialize(int width, int height, int parentWidth, int parentHeight) {
        super.initialize(width, height, parentWidth, parentHeight);
        Log.d(TAG, "width: " + width);
        Log.d(TAG, "height: " + height);
        Log.d(TAG, "pwidth: " + parentWidth);
        Log.d(TAG, "pheight: " + parentHeight);
        setDuration(2500);
        setFillAfter(true);
        setInterpolator(new LinearInterpolator());
    }

    @Override
    protected void applyTransformation(float interpolatedTime, Transformation t) {
        applyTransformationNew(interpolatedTime, t);
    }

    protected void applyTransformationNew(float interpolatedTime, Transformation t) {
        Log.d(TAG, "applyTransformationNew=>transform: " + interpolatedTime);
        final Matrix matrix = t.getMatrix();
        matrix.setScale(interpolatedTime, interpolatedTime);
        matrix.preTranslate(-centerX,-centerY);
        matrix.postTranslate(centerX, centerY);
    }
}

File: Chapter16/SampleViewAnimation/Kotlin/SampleViewAnimation/app/src/main/java/com/ai/android/sampleviewanimation/ViewAnimation.kt
package com.ai.android.sampleviewanimation

import android.util.Log
import android.view.animation.Animation
import android.view.animation.LinearInterpolator
import android.view.animation.Transformation

class ViewAnimation(
    private val centerX: Float,
    private val centerY: Float
): Animation() {

    override fun initialize(width: Int, height: Int, parentWidth: Int, parentHeight: Int) {
        super.initialize(width, height, parentWidth, parentHeight)
        Log.d(TAG, "width: $width")
        Log.d(TAG, "height: $height")
        Log.d(TAG, "pwidth: $parentWidth")
        Log.d(TAG, "pheight: $parentHeight")
        duration = 2500
        fillAfter = true
        interpolator = LinearInterpolator()
    }

    override fun applyTransformation(interpolatedTime: Float, t: Transformation?) {
        applyTransformationNew(interpolatedTime, t)
    }

    private fun applyTransformationNew(interpolatedTime: Float, t: Transformation?) {
        Log.d(TAG, "transform: $interpolatedTime")
        t?.apply {
            matrix?.apply {
                setScale(interpolatedTime, interpolatedTime)
                preTranslate(-centerX, -centerY)
                postTranslate(centerX, centerX)
            }
        }
    }

    companion object {
        const val TAG = "ViewAnimation"
    }
}